votor/src/consensus_pool_service.rs

```diff
@@ -0,0 +1,731 @@
+        let (new_finalized_slot, new_certificates_to_send) =
+            Self::add_message_and_maybe_update_commitment(
+                &root_bank,
+                my_pubkey,
+                &ctx.my_vote_pubkey,
+                message,
+                consensus_pool,
+                events,
+                &ctx.commitment_sender,
+            )?;
+        Self::maybe_update_root_and_send_new_certificates(
+            consensus_pool,
+            &ctx.sharable_banks,
+            &ctx.bls_sender,
+            new_finalized_slot,
+            new_certificates_to_send,
+            standstill_timer,
+            stats,
+        )
+    }
+
+    fn handle_channel_disconnected(
+        ctx: &mut ConsensusPoolContext,
+        channel_name: &str,
+    ) -> Result<(), ()> {
+        info!(
+            "{}: {} disconnected. Exiting",
+            ctx.cluster_info.id(),
+            channel_name
+        );
+        ctx.exit.store(true, Ordering::Relaxed);
+        Err(())
+    }
+
+    // Main loop for the certificate pool service, it only exits when any channel is disconnected
+    fn consensus_pool_ingest_loop(mut ctx: ConsensusPoolContext) -> Result<(), ()> {
+        let mut events = vec![];
+        let mut my_pubkey = ctx.cluster_info.id();
+        let root_bank = ctx.sharable_banks.root();
+        let mut consensus_pool = ConsensusPool::new_from_root_bank(my_pubkey, &root_bank);
+
+        // Wait until migration has completed
+        info!("{}: Certificate pool loop initialized", &my_pubkey);
+        Votor::wait_for_migration_or_exit(&ctx.exit, &ctx.start);
+        info!("{}: Certificate pool loop starting", &my_pubkey);
+        let mut stats = ConsensusPoolServiceStats::new();
+
+        // Standstill tracking
+        let mut standstill_timer = Instant::now();
+
+        // Kick off parent ready
+        let root_block = (root_bank.slot(), root_bank.block_id().unwrap_or_default());
+        let mut highest_parent_ready = root_bank.slot();
+        events.push(VotorEvent::ParentReady {
+            slot: root_bank.slot().checked_add(1).unwrap(),
+            parent_block: root_block,
+        });
+
+        // Ingest votes into certificate pool and notify voting loop of new events
+        while !ctx.exit.load(Ordering::Relaxed) {
+            // Update the current pubkey if it has changed
+            let new_pubkey = ctx.cluster_info.id();
+            if my_pubkey != new_pubkey {
+                my_pubkey = new_pubkey;
+                consensus_pool.update_pubkey(my_pubkey);
+                warn!("Certificate pool pubkey updated to {my_pubkey}");
+            }
+
+            Self::add_produce_block_event(
+                &mut highest_parent_ready,
+                &consensus_pool,
+                &my_pubkey,
+                &mut ctx,
+                &mut events,
+                &mut stats,
+            );
+
+            if standstill_timer.elapsed() > DELTA_STANDSTILL {
```

Comment:
maybe this `DELTA_STANDSTILL` should be provided as an input as part of `ConsensusPoolContext`. This could allow tests to configure this to their liking. I'm guessing we will have some decent number of tests that don't want a 10ms standstill